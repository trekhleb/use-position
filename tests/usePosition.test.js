import React from 'react';
import { Demo } from '../demo/Demo';
import renderer, { act } from 'react-test-renderer';

const mockPosition = {
  coords: {
    latitude: 52.3172414,
    longitude: 4.8717809,
    accuracy: 24,
    heading: 90,
    speed: 0,
  },
  timestamp: 1561815013194,
};

const mockWatcherId = 1;

const mockGeolocation = {
  watchPosition: onChange => {
    onChange(mockPosition);
    return mockWatcherId;
  },
  getCurrentPosition: onChange => onChange(mockPosition),
  clearWatch: jest.fn(),
};

const mockGeolocationError = {
  watchPosition: (onChange, onError) => onError(new Error('User denied Geolocation')),
  getCurrentPosition: (onChange, onError) => onError(new Error('User denied Geolocation')),
  clearWatch: jest.fn(),
};

describe('usePosition', () => {
  it('should return empty values by default', () => {
    const tree = renderer.create(<Demo />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return empty values by default in watching mode', () => {
    const tree = renderer.create(<Demo watch />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return latitude and longitude while watching', () => {
    global.navigator.geolocation = mockGeolocation;
    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo watch />);
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return latitude and longitude while fetching', () => {
    global.navigator.geolocation = mockGeolocation;
    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo />);
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return error while watching', () => {
    global.navigator.geolocation = mockGeolocationError;
    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo watch />);
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return error while fetching', () => {
    global.navigator.geolocation = mockGeolocationError;
    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo />);
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return error if navigator is not supported', () => {
    global.navigator.geolocation = null;
    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo />);
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should clear watcher on component unmount', () => {
    global.navigator.geolocation = mockGeolocation;
    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo watch />);
    });
    act(() => {
      testRenderer.unmount();
    });
    expect(global.navigator.geolocation.clearWatch).toHaveBeenCalledWith(mockWatcherId);
  });

  it('should not clear watcher on component unmount', () => {
    global.navigator.geolocation = mockGeolocation;
    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo />);
    });
    act(() => {
      testRenderer.unmount();
    });
    expect(global.navigator.geolocation.clearWatch).not.toHaveBeenCalled();
  });

  it('should return latitude and longitude even if watch is undefined', () => {
    global.navigator.geolocation = mockGeolocation;
    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo watch={undefined} />);
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not call geolocation when skipped', () => {
    global.navigator.geolocation = {
      watchPosition: jest.fn(),
      getCurrentPosition: jest.fn(),
      clearWatch: jest.fn(),
    };

    let testRenderer;
    act(() => {
      testRenderer = renderer.create(<Demo skip />);
    });
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();

    expect(global.navigator.geolocation.watchPosition).not.toHaveBeenCalled();
    expect(global.navigator.geolocation.getCurrentPosition).not.toHaveBeenCalled();
  });
});
