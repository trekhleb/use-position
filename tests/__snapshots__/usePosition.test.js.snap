// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePosition should not call geolocation when skipped 1`] = `
Array [
  <div>
    Trying to fetch location...
  </div>,
  <br />,
  <code>
    latitude: 
    <br />
    longitude: 
    <br />
    timestamp: 
    <br />
    accuracy: 
    <br />
    speed: 
    <br />
    heading: 
    <br />
    error: 
  </code>,
]
`;

exports[`usePosition should return empty values by default 1`] = `
Array [
  <div>
    Trying to fetch location...
  </div>,
  <br />,
  <code>
    latitude: 
    <br />
    longitude: 
    <br />
    timestamp: 
    <br />
    accuracy: 
    <br />
    speed: 
    <br />
    heading: 
    <br />
    error: 
  </code>,
]
`;

exports[`usePosition should return empty values by default in watching mode 1`] = `
<code>
  latitude: 
  <br />
  longitude: 
  <br />
  timestamp: 
  <br />
  accuracy: 
  <br />
  speed: 
  <br />
  heading: 
  <br />
  error: 
  Geolocation is not supported
</code>
`;

exports[`usePosition should return error if navigator is not supported 1`] = `
<code>
  latitude: 
  <br />
  longitude: 
  <br />
  timestamp: 
  <br />
  accuracy: 
  <br />
  speed: 
  <br />
  heading: 
  <br />
  error: 
  Geolocation is not supported
</code>
`;

exports[`usePosition should return error while fetching 1`] = `
<code>
  latitude: 
  <br />
  longitude: 
  <br />
  timestamp: 
  <br />
  accuracy: 
  <br />
  speed: 
  <br />
  heading: 
  <br />
  error: 
  User denied Geolocation
</code>
`;

exports[`usePosition should return error while watching 1`] = `
<code>
  latitude: 
  <br />
  longitude: 
  <br />
  timestamp: 
  <br />
  accuracy: 
  <br />
  speed: 
  <br />
  heading: 
  <br />
  error: 
  User denied Geolocation
</code>
`;

exports[`usePosition should return latitude and longitude even if watch is undefined 1`] = `
<code>
  latitude: 
  52.3172414
  <br />
  longitude: 
  4.8717809
  <br />
  timestamp: 
  1561815013194
  <br />
  accuracy: 
  24 meters
  <br />
  speed: 
  0
  <br />
  heading: 
  90 degrees
  <br />
  error: 
</code>
`;

exports[`usePosition should return latitude and longitude while fetching 1`] = `
<code>
  latitude: 
  52.3172414
  <br />
  longitude: 
  4.8717809
  <br />
  timestamp: 
  1561815013194
  <br />
  accuracy: 
  24 meters
  <br />
  speed: 
  0
  <br />
  heading: 
  90 degrees
  <br />
  error: 
</code>
`;

exports[`usePosition should return latitude and longitude while watching 1`] = `
<code>
  latitude: 
  52.3172414
  <br />
  longitude: 
  4.8717809
  <br />
  timestamp: 
  1561815013194
  <br />
  accuracy: 
  24 meters
  <br />
  speed: 
  0
  <br />
  heading: 
  90 degrees
  <br />
  error: 
</code>
`;
